cpp/src/arrow/testing/random.cc

```diff
@@ -793,7 +793,7 @@ std::shared_ptr<Array> RandomArrayGenerator::ArrayOf(const Field& field, int64_t
                 values_length, alignment, memory_pool);                              \
     const auto offsets = OffsetsFromLengthsArray(lengths.get(), force_empty_nulls,   \
                                                  alignment, memory_pool);            \
-    return *ARRAY_TYPE::FromArrays(*offsets, *values);                               \
+    return *ARRAY_TYPE::FromArrays(field.type(), *offsets, *values);                 \
```

Comment:
Does map type have similar issue?

Reply:
No, the problem is a bit tricky. Let me explain it:
1. Arrow has `DataType`, `DataType` doesn't including some nullable info itself. `Field` has nullable info.
2. For `DataType` , it child would be array of `Field`, so DataType has children's nullable info

Here if doesn't pass `field.type()` in ctor, it will mark child as nullable. And Map's constructor already marks it.

Reply:
Specifically, if you look at the case for MapType, it already passes the type. So this is just bringing ListType into alignment.

Reply:
Yes, this patch brings `List` and `Struct` into alignment. For testing, I also tests map